File: c++/a03_max_subarray.cc
// Author: Susam Pal
// License: MIT


#include <iostream>
#include <vector>
#include <tuple>
#include <stdexcept>


std::tuple<std::size_t, std::size_t, int>
find_max_crossing_subarray(std::vector<int>& A,std::size_t low,
                           std::size_t mid, std::size_t high)
{
    if (!(low <= mid && mid < high)) {
        throw std::invalid_argument("required low <= mid < high");
    }

    std::size_t max_left;
    auto left_sum = 0;
    auto sum = 0;

    for (auto i = mid; i >= low; i--) {
        sum += A[i];
        if (i == mid || sum > left_sum) {
            left_sum = sum;
            max_left = i;
        }

        if (i == 0)
            break;
    }

    std::size_t max_right;
    auto right_sum = 0;
    sum = 0;

    for (auto j = mid + 1; j <= high; j++) {
        sum += A[j];
        if (j == mid + 1 || sum > right_sum) {
            right_sum = sum;
            max_right = j;
        }
    }

    return std::make_tuple(max_left, max_right, left_sum + right_sum);
}


std::tuple<std::size_t, std::size_t, int>
find_max_subarray(std::vector<int>& A, std::size_t low, std::size_t high)
{
    if (high == low)
        return std::make_tuple(low, high, A[low]);

    auto mid = (low + high) / 2;

    std::size_t left_low, right_low, cross_low;
    std::size_t left_high, right_high, cross_high;
    int left_sum, right_sum, cross_sum;

    std::tie(left_low, left_high, left_sum) =
            find_max_subarray(A, low, mid);

    std::tie(right_low, right_high, right_sum) =
            find_max_subarray(A, mid + 1, high);

    std::tie(cross_low, cross_high, cross_sum) =
            find_max_crossing_subarray(A, low, mid, high);

    if (left_sum >= right_sum && left_sum >= cross_sum)
        return std::make_tuple(left_low, left_high, left_sum);

    else if (right_sum >= left_sum && right_sum >= cross_sum)
        return std::make_tuple(right_low, right_high, right_sum);

    else
        return std::make_tuple(cross_low, cross_high, cross_sum);
}


#ifndef TEST
int main()
{
    std::vector<int> A = {13, -3, -25, 20,  -3, -16, -23, 18,
                          20, -7,  12, -5, -22,  15,  -4,  7};

    std::size_t low;
    std::size_t high;
    int max_sum;

    std::tie(low, high, max_sum) = find_max_subarray(A, 0, A.size() - 1);
    std::cout << low << ' ' << high << ' ' << max_sum << '\n';
}
#endif

File: python/a03_max_subarray.py
#!/usr/bin/env python3

# Author: Sunaina Pai
# License: MIT


def find_max_crossing_subarray(A, low, mid, high):
    assert low <= mid < high

    sum_ = 0

    for i in range(mid, low - 1, -1):
        sum_ += A[i]
        if i == mid or sum_ > left_sum:
            left_sum = sum_
            max_left = i

    sum_ = 0

    for j in range(mid + 1, high + 1):
        sum_ += A[j]
        if j == mid + 1 or sum_ > right_sum:
            right_sum = sum_
            max_right = j

    return max_left, max_right, left_sum + right_sum


def find_max_subarray(A, low, high):
    if low == high:
        return low, high, A[low]

    mid = (low + high) // 2

    left_low, left_high, left_sum = (
            find_max_subarray(A, low, mid))

    right_low, right_high, right_sum = (
            find_max_subarray(A, mid + 1, high))

    cross_low, cross_high, cross_sum = (
            find_max_crossing_subarray(A, low, mid, high))

    if left_sum >= right_sum and left_sum >= cross_sum:
        return left_low, left_high, left_sum

    if right_sum >= left_sum and right_sum >= cross_sum:
        return right_low, right_high, right_sum

    return cross_low, cross_high, cross_sum


if __name__ == '__main__':
    A = [13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7]
    low, high, max_sum = find_max_subarray(A, 0, len(A) - 1)
    print(low, high, max_sum)
